File: web/documentserver-example/python/src/format/format.py
#
# (c) Copyright Ascensio System SIA 2023
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

from pathlib import Path
from msgspec.json import decode
from msgspec import Struct
from src.memoize import memoize

class Format(Struct):
    name: str
    type: str
    actions: list[str]
    convert: list[str]
    mime: list[str]

    def extension(self) -> str:
        return f'.{self.name}'

class FormatManager():
    def fillable_extensions(self) -> list[str]:
        formats = self.fillable()
        mapped = map(lambda format: format.extension(), formats)
        return list(mapped)

    def fillable(self) -> list[Format]:
        formats = self.all()
        filtered = filter(lambda format: 'fill' in format.actions, formats)
        return list(filtered)

    def viewable_extensions(self) -> list[str]:
        formats = self.viewable()
        mapped = map(lambda format: format.extension(), formats)
        return list(mapped)

    def viewable(self) -> list[Format]:
        formats = self.all()
        filtered = filter(lambda format: 'view' in format.actions, formats)
        return list(filtered)

    def editable_extensions(self) -> list[str]:
        formats = self.editable()
        mapped = map(lambda format: format.extension(), formats)
        return list(mapped)

    def editable(self) -> list[Format]:
        formats = self.all()
        filtered = filter(
            lambda format: (
                'edit' in format.actions or
                'lossy-edit' in format.actions
            ),
            formats
        )
        return list(filtered)

    def convertible_extensions(self) -> list[str]:
        formats = self.convertible()
        mapped = map(lambda format: format.extension(), formats)
        return list(mapped)

    def convertible(self) -> list[Format]:
        formats = self.all()
        filtered = filter(
            lambda format: (
                format.type == 'cell' and 'xlsx' in format.convert or
                format.type == 'slide' and 'pptx' in format.convert or
                format.type == 'word' and 'docx' in format.convert
            ),
            formats
        )
        return list(filtered)

    def spreadsheet_extensions(self) -> list[str]:
        formats = self.spreadsheets()
        mapped = map(lambda format: format.extension(), formats)
        return list(mapped)

    def spreadsheets(self) -> list[Format]:
        formats = self.all()
        filtered = filter(lambda format: format.type == 'cell', formats)
        return list(filtered)

    def presentation_extensions(self) -> list[str]:
        formats = self.presentations()
        mapped = map(lambda format: format.extension(), formats)
        return list(mapped)

    def presentations(self) -> list[Format]:
        formats = self.all()
        filtered = filter(lambda format: format.type == 'slide', formats)
        return list(filtered)

    def document_extensions(self) -> list[str]:
        formats = self.documents()
        mapped = map(lambda format: format.extension(), formats)
        return list(mapped)

    def documents(self) -> list[Format]:
        formats = self.all()
        filtered = filter(lambda format: format.type == 'word', formats)
        return list(filtered)

    def all_extensions(self) -> list[str]:
        formats = self.all()
        mapped = map(lambda format: format.extension(), formats)
        return list(mapped)

    @memoize
    def all(self) -> list[Format]:
        path = self.__file()
        with open(path, 'r', encoding='utf-8') as file:
            contents = file.read()
            return decode(contents, type=list[Format])

    def __file(self) -> Path:
        directory = self.__directory()
        return directory.joinpath('onlyoffice-docs-formats.json')

    def __directory(self) -> Path:
        current_file = Path(__file__)
        directory = current_file.joinpath(
            '..',
            '..',
            '..',
            'assets',
            'document-formats'
        )
        return directory.resolve()

File: web/documentserver-example/ruby/app/format/format.rb
#
# (c) Copyright Ascensio System SIA 2025
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# frozen_string_literal: true
# typed: true

require 'pathname'
require 'sorbet-runtime'

# Struct representing a document format with properties.
class Format < T::Struct
  extend T::Sig

  const :name, String
  const :type, String
  const :actions, T::Array[String]
  const :convert, T::Array[String]
  const :mime, T::Array[String]

  sig { params(hash: T.untyped, strict: T.untyped).returns(Format) }
  def self.from_hash(hash, strict = nil)
    super
  end

  sig { returns(String) }
  def extension
    ".#{name}"
  end
end

# FormatManager is responsible for managing document formats and providing various lists of supported extensions.
class FormatManager
  extend T::Sig

  sig { returns(T::Array[String]) }
  def fillable_extensions
    fillable.map(&:extension)
  end

  sig { returns(T::Array[Format]) }
  def fillable
    all.filter do |format|
      format.actions.include?('fill')
    end
  end

  sig { returns(T::Array[String]) }
  def viewable_extensions
    viewable.map(&:extension)
  end

  sig { returns(T::Array[Format]) }
  def viewable
    all.filter do |format|
      format.actions.include?('view')
    end
  end

  sig { returns(T::Array[String]) }
  def editable_extensions
    editable.map(&:extension)
  end

  sig { returns(T::Array[Format]) }
  def editable
    all.filter do |format|
      format.actions.include?('edit') ||
        format.actions.include?('lossy-edit')
    end
  end

  sig { returns(T::Array[String]) }
  def convertible_extensions
    convertible.map(&:extension)
  end

  sig { returns(T::Array[Format]) }
  def convertible
    all.filter do |format|
      format.actions.include?('auto-convert')
    end
  end

  sig { returns(T::Array[String]) }
  def diagram_extensions
    diagram.map(&:extension)
  end

  sig { returns(T::Array[Format]) }
  def diagram
    all.filter do |format|
      format.type == 'diagram'
    end
  end

  sig { returns(T::Array[String]) }
  def pdf_extensions
    pdf.map(&:extension)
  end

  sig { returns(T::Array[Format]) }
  def pdf
    all.filter do |format|
      format.type == 'pdf'
    end
  end

  sig { returns(T::Array[String]) }
  def spreadsheet_extensinons
    spreadsheets.map(&:extension)
  end

  sig { returns(T::Array[Format]) }
  def spreadsheets
    all.filter do |format|
      format.type == 'cell'
    end
  end

  sig { returns(T::Array[String]) }
  def presentation_extensinons
    presentations.map(&:extension)
  end

  sig { returns(T::Array[Format]) }
  def presentations
    all.filter do |format|
      format.type == 'slide'
    end
  end

  sig { returns(T::Array[String]) }
  def document_extensinons
    documents.map(&:extension)
  end

  sig { returns(T::Array[Format]) }
  def documents
    all.filter do |format|
      format.type == 'word'
    end
  end

  sig { returns(T::Array[String]) }
  def all_extensions
    all.map(&:extension)
  end

  sig { returns(T::Array[Format]) }
  def all
    return @all if defined?(@all)

    content = file.read
    hash = JSON.parse(content)
    @all ||= hash.map do |item|
      Format.from_hash(item)
    end
  end

  sig { returns(Pathname) }
  private def file
    directory.join('onlyoffice-docs-formats.json')
  end

  sig { returns(Pathname) }
  private def directory
    current_directory = Pathname(T.must(__dir__))
    directory = current_directory.join('..', '..', 'assets', 'document-formats')
    directory.cleanpath
  end
end
